sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,125 @@
+#include <sys/devfs.h>
+#include <sys/vnode.h>
+#include <sys/linker_set.h>
+#include <sys/queue.h>
+#include <sys/proc.h>
+#include <sys/thread.h>
+#include <sys/klog.h>
+#include <sys/malloc.h>
+#include <stdio.h>
+
+/* Implementation of /dev/procstat
+ *
+ * When /dev/procstat is opened it takes a snapshot of info of all existing
+ * processes.
+ *
+ * During read calls to /dev/procstat this info can be read.
+ *
+ * Example:
+ * euid   pid    ppid    pgrp    state    td_name  elfpath
+ * 0       1       0       1       R       init    /bin/ksh
+ *
+ */
+
+#define PSTAT_SIZE 1024
+#define BUF_SIZE (PSTAT_SIZE - sizeof(int))
+
+static char state[4] = {
+  [PS_NORMAL] = 'R',
+  [PS_STOPPED] = 'S',
+  [PS_DYING] = 'D',
+  [PS_ZOMBIE] = 'Z',
+};
+
+typedef struct pstat_data {
+  size_t len;
+  char buf[];
+} pstat_t;
+
+static int dev_procstat_read(file_t *f, uio_t *uio) {
+  pstat_t *pst = f->f_data;
+
+  if ((size_t)f->f_offset >= pst->len)
+    return 0;
+
+  uio->uio_offset = f->f_offset;
+  size_t len = MIN(uio->uio_resid, pst->len);
+  int error = uiomove(pst->buf, len, uio);
+
+  f->f_offset = uio->uio_offset;
+  return error;
+}
+
+static fileops_t dev_procstat_fileops = {
+  .fo_read = dev_procstat_read,
+  .fo_close = default_vnclose,
+  .fo_stat = default_vnstat,
+  .fo_ioctl = default_vnioctl,
+};
+
+static int write_to_buf(char *buf, size_t max_len, proc_t *p) {
+  if (p->p_pid == 0)
+    return 0;
```

Comment:
Is it possible?

Reply:
Yes, there exists `proc0`. I think we don't want to show it on list of processes.

Reply:
Okay... Now I understand. It wasn't obvious for me to put that logic inside `write_to_buf`. From my point of view that function should do one simple thing - writing data into buffer, but do whatever is more clean solution for you.